euphoria-core/src/main/java/cz/seznam/euphoria/core/util/PrefetchIterator.java

```diff
@@ -0,0 +1,65 @@
+/**
+ * Copyright 2016-2017 Seznam.cz, a.s.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.seznam.euphoria.core.util;
+
+import java.util.Iterator;
+import java.util.stream.Stream;
+
+/**
+ * Iterator wrapper allowing to look at next value using {@link #prefetch()}
+ * without moving the iterator forward, so following call of {@link #next()}
+ * will return the same value.
+ */
+public class PrefetchIterator<T> implements Iterator<T> {
+  private final Iterator<T> it;
+  private T prefetched = null;
+
+  public PrefetchIterator(Iterator<T> src) {
+    it = src;
+  }
+  public PrefetchIterator(Stream<T> src) {
+    this(src.iterator());
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public boolean hasNext() {
+    return prefetched != null || it.hasNext();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public T next() {
+    if (prefetched != null) {
+      T ret = prefetched;
+      prefetched = null;
+      return ret;
+    } else {
+      return it.next();
+    }
+  }
+
+  /**
+   * Get next value without moving the iterator forward.
+   *
+   * @return Next value (or null on the end)
+   */
+  public T prefetch() {
```

Comment:
Can be name this `peek`? That is compliant with naming conventions of other methods performing this task.

Reply:
I'm in doubt about `peek()`.

It has different meaning in java 8 stream API and maybe someone would like to add `PeekingStream` later doing the same thing on stream.

Reply:
I'd say, that the naming of `Stream#peek` seems unfortunate, but OK. Can we come up with some other naming that would be clear to users? I have a strong feeling that the current naming is not self-explanatory.

Reply:
OK, more names:

- `LookupIterator::lookup()`
- `PeekingIterator::peek()`
- `PeepIterator::peep()`
- `SurveyIterator::survey()`
- `InspectIterator::inspect()`
- ?

Reply:
I still vote for `peek`, but maybe others might have some opinion on this.

Reply:
@dmvk Cool, I wasn't aware of that. I think `Iterators.peekingIterator(values.iterator())` does exactly this and we shouldn't reinvent the wheel. Also, as discussed over slack, motivation for this is probably missing #131 
We should probably close this.